src/derived.rs

```diff
@@ -501,6 +515,37 @@ where
     }
 }
 
+struct PanicGuard<'db, DB, Q>
+where
+    DB: Database + 'db,
+    Q: QueryFunction<DB>,
+{
+    map: &'db RwLock<FxHashMap<Q::Key, QueryState<DB, Q>>>,
+    key: &'db Q::Key,
+}
+
+impl<'db, DB, Q> PanicGuard<'db, DB, Q>
+where
+    DB: Database + 'db,
+    Q: QueryFunction<DB>,
+{
+    fn new(map: &'db RwLock<FxHashMap<Q::Key, QueryState<DB, Q>>>, key: &'db Q::Key) -> Self {
+        Self { map, key }
+    }
+}
+
+impl<'db, DB, Q> Drop for PanicGuard<'db, DB, Q>
+where
+    DB: Database + 'db,
+    Q: QueryFunction<DB>,
+{
+    fn drop(&mut self) {
+        let map = self.map.upgradable_read();
```

Comment:
Here, you might as well just do `let map = self.map.write()` and not bother with the "upgradable read" and the "upgrade". (A read-write lock has two modes: read (like shared) and write (one write, no readers) -- an "upgradable read" allows readers but permits you to "upgrade" to a write-lock atomically with respect to any writes -- i.e., there can only be one "upgradable reader" at a time.)